netbench-orchestrator/src/ec2_utils.rs

```diff
@@ -1,21 +1,149 @@
+use tracing::{debug, error, info};
 
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
-pub struct PubIp(pub IpAddr);
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
-pub struct PrivIp(pub IpAddr);
+mod instance;
+mod launch_plan;
+mod networking;
+mod types;
 
-impl std::fmt::Display for PrivIp {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.0)
+pub use types::{Az, PrivIp};
+
+const RETRY_COUNT: usize = 25;
+const RETRY_BACKOFF: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+pub struct InfraDetail {
+    pub security_group_id: String,
+    pub clients: Vec<InstanceDetail>,
+    pub servers: Vec<InstanceDetail>,
+    placement_map: HashMap<Az, PlacementGroup>,
+}
+
+impl InfraDetail {
+    pub async fn cleanup(&self, ec2_client: &aws_sdk_ec2::Client) -> OrchResult<()> {
+        // instances must be deleted before other infra
+        self.delete_instances(ec2_client).await?;
+
+        self.delete_placement_group(ec2_client).await?;
+        // generally takes a long time so attempt this last
+        self.delete_security_group(ec2_client).await?;
+        Ok(())
+    }
+
+    pub fn public_server_ips(&self) -> Vec<&PubIp> {
+        self.servers
+            .iter()
+            .map(|instance| instance.host_ips().public_ip())
+            .collect()
+    }
+
+    pub fn private_server_ips(&self) -> Vec<&PrivIp> {
+        self.servers
+            .iter()
+            .map(|instance| instance.host_ips().private_ip())
+            .collect()
+    }
+
+    pub fn public_client_ips(&self) -> Vec<&PubIp> {
+        self.clients
+            .iter()
+            .map(|instance| instance.host_ips().public_ip())
+            .collect()
     }
 }
 
-impl std::fmt::Display for PubIp {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.0)
+impl InfraDetail {
+    async fn delete_instances(&self, ec2_client: &aws_sdk_ec2::Client) -> OrchResult<()> {
+        info!("Start: deleting instances");
+        let ids: Vec<String> = self
+            .servers
+            .iter()
+            .chain(self.clients.iter())
+            .map(|instance| instance.instance_id().to_string())
+            .collect();
+
+        ec2_client
+            .terminate_instances()
+            .set_instance_ids(Some(ids))
+            .send()
+            .await
+            .map_err(|err| OrchError::Ec2 {
+                dbg: err.to_string(),
+            })?;
+
+        Ok(())
+    }
+
+    async fn delete_security_group(&self, ec2_client: &aws_sdk_ec2::Client) -> OrchResult<()> {
+        info!("Start: deleting security groups");
+
+        let mut deleted_sec_group = ec2_client
+            .delete_security_group()
+            .group_id(self.security_group_id.to_string())
+            .send()
+            .await;
+
+        let mut retries = RETRY_COUNT;
+        while deleted_sec_group.is_err() && retries > 0 {
```

Comment:
only match on [`DependencyViolation`](https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/operation/delete_security_group/builders/struct.DeleteSecurityGroupFluentBuilder.html#method.send)

Reply:
Also can use https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/operation/describe_instances/builders/struct.DescribeInstancesFluentBuilder.html to check instance status.

Reply:
Change implemented in https://github.com/aws/s2n-netbench/pull/26/commits/ef5a2abdc218bfa4456c992b70c4e302fa1f1c82